geode-core/src/distributedTest/java/org/apache/geode/internal/cache/tier/sockets/RedundancyLevelPart1DUnitTest.java

```diff
@@ -296,25 +303,243 @@ public void testRedundancySpecifiedNonPrimaryEPFailsDetectionByUnregisterInteres
+    Region<Object, Object> region = cache.getRegion(REGION_NAME);
+    assertThat(region).isNotNull();
+    List<String> list = new ArrayList<>();
+    list.add(K1);
+    list.add(K2);
+    region.registerInterest(list, InterestResultPolicy.KEYS_VALUES);
+  }
+
+  private static void unregisterInterest() {
+    Region<String, String> r = cache.getRegion(Region.SEPARATOR + REGION_NAME);
+    r.unregisterInterest("k1");
+  }
+
+  private void stopServer() {
+    Iterator<CacheServer> iterator = cache.getCacheServers().iterator();
+    if (iterator.hasNext()) {
+      CacheServer server = iterator.next();
+      server.stop();
+    }
+  }
+
+  private void createClientCache(int redundancy, int socketReadTimeout, int retryInterval) {
+    if (!failOverDetectionByCCU.get()) {
+      clientServerObserver =
+          ClientServerObserverHolder.setInstance(new ClientServerObserverAdapter() {
+            @Override
+            public void beforeFailoverByCacheClientUpdater(ServerLocation epFailed) {
+              try {
+                latch.await(TIMEOUT_MILLIS, MILLISECONDS);
+              } catch (InterruptedException ie) {
+                // expected - test will shut down the cache which will interrupt
+                // the CacheClientUpdater thread that invoked this method
+                Thread.currentThread().interrupt();
+              }
+            }
+          });
+    }
+
+    Properties props = new Properties();
+    props.setProperty(MCAST_PORT, "0");
+    props.setProperty(LOCATORS, "");
+
+    cache = cacheRule.getOrCreateCache(props);
+
+    pool = (PoolImpl) PoolManager.createFactory()
+        .addServer(hostname, port0)
+        .addServer(hostname, port1)
+        .addServer(hostname, port2)
+        .addServer(hostname, port3)
+        .setSubscriptionEnabled(true)
+        .setReadTimeout(socketReadTimeout)
+        .setSocketBufferSize(32768)
+        .setMinConnections(8)
+        .setSubscriptionRedundancy(redundancy)
+        .setRetryAttempts(5)
+        .setPingInterval(retryInterval)
+        .create("DurableClientReconnectDUnitTestPool");
+
+    cache.createRegionFactory()
+        .setScope(Scope.DISTRIBUTED_ACK)
+        .setPoolName(pool.getName())
+        .create(REGION_NAME);
+
+    createEntriesK1andK2();
+    registerK1AndK2();
+  }
+
+  private int createServerCache() throws Exception {
+    cache = cacheRule.getOrCreateCache();
+
+    cache.createRegionFactory(RegionShortcut.REPLICATE).setEnableSubscriptionConflation(true)
+        .create(REGION_NAME);
+
+    CacheServer cacheServer = cache.addCacheServer();
+
+    cacheServer.setMaximumTimeBetweenPings(180000);
+    cacheServer.setPort(0);
+    cacheServer.start();
+
+    return cacheServer.getPort();
+  }
+
+  private void verifyInterestRegistration() {
+    await().untilAsserted(() -> assertThat(cache.getCacheServers()).hasSize(1));
+
+    CacheServerImpl cacheServer = (CacheServerImpl) cache.getCacheServers().iterator().next();
```

Comment:
Another good thing to change is to move away from direct usage of implementation classes towards usage of interfaces. You'll find that this works:
```
InternalCacheServer cacheServer = (InternalCacheServer) cache.getCacheServers().iterator().next();
```

Reply:
Done. Thank you.